server/CGameHandler.cpp

```diff
@@ -3931,6 +3931,9 @@ void CGameHandler::castSpell(const spells::Caster * caster, SpellID spellID, con
 
 	const CSpell * s = spellID.toSpell();
 	s->adventureCast(spellEnv, p);
+
+	if(const auto hero = caster->getHeroCaster())
+		verifyChargedArtifactUsed(hero->id, spellID);
```

Comment:
@SoundSSGood 
I am talking about this place.
Perhaps that check for "hero has spell in spellbook" should be done here instead?
And only call `verifyChargedArtifactUsed` if hero does not have that spell in spellbook and has to expend charge from artifact

Reply:
Theoretically, there could still be artifacts without charges, with a bonus in the form of a spell. Or temporary bonuses. I thought it would be more convenient to move the check to a separate method.

Reply:
Like scrolls? In that case - yeah, probably makes sense.